In [1]:
import sys
print(sys.version)
#For this notebook to work, Python must be 3.6.4 or 3.6.5

import numpy as np
import pandas as pd
from IPython.display import display

from plotnine import *
import matplotlib.pyplot as plt
import seaborn as sns

3.6.5 |Anaconda, Inc.| (default, Mar 29 2018, 13:32:41) [MSC v.1900 64 bit (AMD64)]


In [2]:
#Import Data Set
cancer = pd.read_csv('data/breast-cancer-wisconsin-data/data.csv', sep=',')

In [3]:
#Inspect data set
display(cancer[:][:5])
print(cancer.shape)

id diagnosis  radius_mean  texture_mean  perimeter_mean  area_mean  \
0    842302         M        17.99         10.38          122.80     1001.0   
1    842517         M        20.57         17.77          132.90     1326.0   
2  84300903         M        19.69         21.25          130.00     1203.0   
3  84348301         M        11.42         20.38           77.58      386.1   
4  84358402         M        20.29         14.34          135.10     1297.0   

   smoothness_mean  compactness_mean  concavity_mean  concave points_mean  \
0          0.11840           0.27760          0.3001              0.14710   
1          0.08474           0.07864          0.0869              0.07017   
2          0.10960           0.15990          0.1974              0.12790   
3          0.14250           0.28390          0.2414              0.10520   
4          0.10030           0.13280          0.1980              0.10430   

   ...  radius_worst  texture_worst  perimeter_worst  area_worst  \
0  ...         25.38          17.33           184.60      2019.0   
1  ...         24.99          23.41           158.80      1956.0   
2  ...         23.57          25.53           152.50      1709.0   
3  ...         14.91          26.50            98.87       567.7   
4  ...         22.54          16.67           152.20      1575.0   

   smoothness_worst  compactness_worst  concavity_worst  concave points_worst  \
0            0.1622             0.6656           0.7119                0.2654   
1            0.1238             0.1866           0.2416                0.1860   
2            0.1444             0.4245           0.4504                0.2430   
3            0.2098             0.8663           0.6869                0.2575   
4            0.1374             0.2050           0.4000                0.1625   

   symmetry_worst  fractal_dimension_worst  
0          0.4601                  0.11890  
1          0.2750                  0.08902  
2          0.3613                  0.08758  
3          0.6638                  0.17300  
4          0.2364                  0.07678  

[5 rows x 32 columns]

(569, 32)


In [4]:
#Check data types 
cancer.dtypes

id                           int64
diagnosis                   object
radius_mean                float64
texture_mean               float64
perimeter_mean             float64
area_mean                  float64
smoothness_mean            float64
compactness_mean           float64
concavity_mean             float64
concave points_mean        float64
symmetry_mean              float64
fractal_dimension_mean     float64
radius_se                  float64
texture_se                 float64
perimeter_se               float64
area_se                    float64
smoothness_se              float64
compactness_se             float64
concavity_se               float64
concave points_se          float64
symmetry_se                float64
fractal_dimension_se       float64
radius_worst               float64
texture_worst              float64
perimeter_worst            float64
area_worst                 float64
smoothness_worst           float64
compactness_worst          float64
concavity_worst     

In [5]:
# Divide the data into X and y (output, labels)
X = cancer.iloc[:, 2:].values
y_categorical = cancer.iloc[:, 1].values

In [6]:
# Sanity check
display(X[:][:5])
print(X.shape)

array([[1.799e+01, 1.038e+01, 1.228e+02, 1.001e+03, 1.184e-01, 2.776e-01,
        3.001e-01, 1.471e-01, 2.419e-01, 7.871e-02, 1.095e+00, 9.053e-01,
        8.589e+00, 1.534e+02, 6.399e-03, 4.904e-02, 5.373e-02, 1.587e-02,
        3.003e-02, 6.193e-03, 2.538e+01, 1.733e+01, 1.846e+02, 2.019e+03,
        1.622e-01, 6.656e-01, 7.119e-01, 2.654e-01, 4.601e-01, 1.189e-01],
       [2.057e+01, 1.777e+01, 1.329e+02, 1.326e+03, 8.474e-02, 7.864e-02,
        8.690e-02, 7.017e-02, 1.812e-01, 5.667e-02, 5.435e-01, 7.339e-01,
        3.398e+00, 7.408e+01, 5.225e-03, 1.308e-02, 1.860e-02, 1.340e-02,
        1.389e-02, 3.532e-03, 2.499e+01, 2.341e+01, 1.588e+02, 1.956e+03,
        1.238e-01, 1.866e-01, 2.416e-01, 1.860e-01, 2.750e-01, 8.902e-02],
       [1.969e+01, 2.125e+01, 1.300e+02, 1.203e+03, 1.096e-01, 1.599e-01,
        1.974e-01, 1.279e-01, 2.069e-01, 5.999e-02, 7.456e-01, 7.869e-01,
        4.585e+00, 9.403e+01, 6.150e-03, 4.006e-02, 3.832e-02, 2.058e-02,
        2.250e-02, 4.571e-03, 2.357e

(569, 30)


In [7]:
# We can see that # 19, 20 and 21 are Benign within the array that goes from 0:29
display(y_categorical[:30])
print(y_categorical.shape)

array(['M', 'M', 'M', 'M', 'M', 'M', 'M', 'M', 'M', 'M', 'M', 'M', 'M',
       'M', 'M', 'M', 'M', 'M', 'M', 'B', 'B', 'B', 'M', 'M', 'M', 'M',
       'M', 'M', 'M', 'M'], dtype=object)

(569,)


In [8]:
# Encoding categorical data
from sklearn.preprocessing import LabelEncoder
labelencoder_X_1 = LabelEncoder()
y = labelencoder_X_1.fit_transform(y_categorical)

In [9]:
# We can see that # 19, 20 and 21 are "0" within the array that goes from 0:29
# Therefore, the sanity check confirms that we have done the encoding correctly
display(y[:30])
print(y.shape)

array([1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 0,
       1, 1, 1, 1, 1, 1, 1, 1], dtype=int64)

(569,)


In [10]:
# Splitting the dataset into the Training set and Test set
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state = 1)

In [11]:
# Due to the nature of ANNs we must scale our features before fitting a Neural Network Model
from sklearn.preprocessing import StandardScaler
sc = StandardScaler()
X_train = sc.fit_transform(X_train)
X_test = sc.transform(X_test)

In [12]:
#Import packages required for an ANN
import keras
from keras.models import Sequential
from keras.layers import Dense, Dropout, Activation

C:\Users\Stephenson\Anaconda3\lib\site-packages\h5py\__init__.py:36: FutureWarning: Conversion of the second argument of issubdtype from `float` to `np.floating` is deprecated. In future, it will be treated as `np.float64 == np.dtype(float).type`.
  from ._conv import register_converters as _register_converters
Using TensorFlow backend.


In [13]:
#Ensure recomputability
from numpy.random import seed
seed(4)
from tensorflow import set_random_seed
set_random_seed(9)

#Create the neural network
ann = Sequential()

#Add input and 1st hidden layer with a dropout of 20% prob
ann.add(Dense(16, activation='relu', kernel_initializer='uniform', input_dim=(30)))
ann.add(Dropout(0.2))
#2nd hidden layer with a dropout of 20% prob
ann.add(Dense(16, activation='relu', kernel_initializer='uniform'))
ann.add(Dropout(0.2))
#Output layer, sigmoid activation function for binary output
ann.add(Dense(1, activation='sigmoid', kernel_initializer='uniform'))

#Compile the model
ann.compile(optimizer='adam', loss='binary_crossentropy', metrics=['accuracy'])
ann.fit(X_train, y_train, epochs=150, batch_size=100)


Instructions for updating:
Colocations handled automatically by placer.
Instructions for updating:
Please use `rate` instead of `keep_prob`. Rate should be set to `rate = 1 - keep_prob`.
Instructions for updating:
Use tf.cast instead.
Epoch 1/150
455/455 [==============================] - 0s 884us/step - loss: 0.6927 - acc: 0.6176
Epoch 2/150
455/455 [==============================] - 0s 64us/step - loss: 0.6909 - acc: 0.6440
Epoch 3/150
455/455 [==============================] - 0s 44us/step - loss: 0.6884 - acc: 0.6484
Epoch 4/150
455/455 [==============================] - 0s 57us/step - loss: 0.6843 - acc: 0.6703
Epoch 5/150
455/455 [==============================] - 0s 46us/step - loss: 0.6778 - acc: 0.7143
Epoch 6/150
455/455 [==============================] - 0s 48us/step - loss: 0.6687 - acc: 0.7560
Epoch 7/150
455/455 [==============================] - 0s 31us/step - loss: 0.6560 - acc: 0.8242
Epoch 8/150
455/455 [==============================] - 0s 42us/step - loss: 0.6372 - 

455/455 [==============================] - 0s 35us/step - loss: 0.0476 - acc: 0.9846
Epoch 76/150
455/455 [==============================] - 0s 73us/step - loss: 0.0402 - acc: 0.9890
Epoch 77/150
455/455 [==============================] - 0s 59us/step - loss: 0.0503 - acc: 0.9890
Epoch 78/150
455/455 [==============================] - 0s 48us/step - loss: 0.0500 - acc: 0.9846
Epoch 79/150
455/455 [==============================] - 0s 35us/step - loss: 0.0491 - acc: 0.9868
Epoch 80/150
455/455 [==============================] - 0s 59us/step - loss: 0.0448 - acc: 0.9890
Epoch 81/150
455/455 [==============================] - 0s 31us/step - loss: 0.0437 - acc: 0.9868
Epoch 82/150
455/455 [==============================] - 0s 26us/step - loss: 0.0534 - acc: 0.9846
Epoch 83/150
455/455 [==============================] - 0s 44us/step - loss: 0.0414 - acc: 0.9868
Epoch 84/150
455/455 [==============================] - 0s 33us/step - loss: 0.0491 - acc: 0.9846
Epoch 85/150
455/455 [===========

In [23]:

prediction=ann.predict(X_test)

y_pred=[1 if x == True else 0 for [x] in prediction >= 0.5]

print("Accuracy score: ",accuracy_score(y_test, y_pred))

Accuracy score:  0.9824561403508771


In [24]:
#Let's tune the drop out parameter
from keras.wrappers.scikit_learn import KerasClassifier

#Create the model for KerasClassifier
def create_model(dropout_rate=0.0):
            ann  = Sequential()
            ann.add(Dense(16, activation='relu', kernel_initializer='uniform', input_dim=(30)))
            ann.add(Dropout(dropout_rate))
            ann.add(Dense(16, activation='relu', kernel_initializer='uniform'))
            ann.add(Dropout(dropout_rate))
            ann.add(Dense(1, activation='sigmoid', kernel_initializer='uniform'))
            ann.compile(optimizer='adam', loss='binary_crossentropy', metrics=['accuracy'])
            return ann

#Set up our pipeline for gridsearchCV
from sklearn.pipeline import Pipeline 
from sklearn.model_selection import GridSearchCV

#Set for Reproducibility 
from numpy.random import seed
seed(9)

#Setting up the estimator for gridsearchcv
model = KerasClassifier(build_fn=create_model, verbose=10)

#Defining paramater grid
#Since accuracy was extremely high for drop_out rate of 20%, we'll select our paramater grid around 0.2
dropout_rate = [0.19, 0.20, 0.21]
param_grid = dict(dropout_rate = dropout_rate)

#Using CV=5
grid_search = GridSearchCV(estimator = model, param_grid = param_grid, cv=5)
#Decrease number of epochs to 10 for computational feasibility
grid_search.fit(X_train, y_train, epochs=10, batch_size=100)
print("Best parameters: {}".format(grid_search.best_params_))
print("Best cross-validation score: {:.4f}".format(grid_search.best_score_))

#Compute mean accuracy scores for each gridsearched parameter
means = grid_search.cv_results_['mean_test_score']
stds = grid_search.cv_results_['std_test_score']
params = grid_search.cv_results_['params']
for mean, stdev, param in zip(means, stds, params):
    print("%f (%f) with: %r" % (mean, stdev, param))

Epoch 1/10
Epoch 2/10
Epoch 3/10
Epoch 4/10
Epoch 5/10
Epoch 6/10
Epoch 7/10
Epoch 8/10
Epoch 9/10
Epoch 10/10
Epoch 1/10
Epoch 2/10
Epoch 3/10
Epoch 4/10
Epoch 5/10
Epoch 6/10
Epoch 7/10
Epoch 8/10
Epoch 9/10
Epoch 10/10
Epoch 1/10
Epoch 2/10
Epoch 3/10
Epoch 4/10
Epoch 5/10
Epoch 6/10
Epoch 7/10
Epoch 8/10
Epoch 9/10
Epoch 10/10
Epoch 1/10
Epoch 2/10
Epoch 3/10
Epoch 4/10
Epoch 5/10
Epoch 6/10
Epoch 7/10
Epoch 8/10
Epoch 9/10
Epoch 10/10
Epoch 1/10
Epoch 2/10
Epoch 3/10
Epoch 4/10
Epoch 5/10
Epoch 6/10
Epoch 7/10
Epoch 8/10
Epoch 9/10
Epoch 10/10
Epoch 1/10
Epoch 2/10
Epoch 3/10
Epoch 4/10
Epoch 5/10
Epoch 6/10
Epoch 7/10
Epoch 8/10
Epoch 9/10
Epoch 10/10
Epoch 1/10
Epoch 2/10
Epoch 3/10
Epoch 4/10
Epoch 5/10
Epoch 6/10
Epoch 7/10
Epoch 8/10
Epoch 9/10
Epoch 10/10
Epoch 1/10
Epoch 2/10
Epoch 3/10
Epoch 4/10
Epoch 5/10
Epoch 6/10
Epoch 7/10
Epoch 8/10
Epoch 9/10
Epoch 10/10
Epoch 1/10
Epoch 2/10
Epoch 3/10
Epoch 4/10
Epoch 5/10
Epoch 6/10
Epoch 7/10
Epoch 8/10
Epoch 9/10
Epoch 10/10
E

In [25]:
# Predicting the Test set results with the best parameters from tuning
prediction = grid_search.predict(np.array(X_test))
print("Test accuracy score: {:.2f}".format(grid_search.score(X_test,y_test)))

Test accuracy score: 0.93


In [26]:
from sklearn.metrics import classification_report, confusion_matrix, accuracy_score

y_pred=[1 if x == True else 0 for [x] in prediction >= 0.5]

#Accuracy: as shown above!
print("Accuracy score: ",accuracy_score(y_test, y_pred))

#Computing Confusion Matrix
confusion = confusion_matrix(y_test, y_pred)
print("Confusion matrix:\n{}".format(confusion))

#Precision, Recall and F1 score 
print(classification_report(y_test, y_pred))

Accuracy score:  0.9298245614035088
Confusion matrix:
[[69  3]
 [ 5 37]]
             precision    recall  f1-score   support

          0       0.93      0.96      0.95        72
          1       0.93      0.88      0.90        42

avg / total       0.93      0.93      0.93       114



#Conclusion#

Write answer here:

#####################################################################################################################

During our initial training stage, the neural network model achieved an impressive accuracy of 0.98 (near perfect) when setting the drop_out rate to be 20%. This shows that our model was able to effectively seperate between the Malignant and Benign diagnosis data points. However, it is important not to be overly joyous at this stage as the result only means that the model is able to generalise well on our training data.

To find the most optimal drop_out parameter for the model, the GridSearchCV process was performed which gave us an algorithm with model parameter of dropout_rate: 0.20 which yielded an accuracy score of 0.93. At first glance, it might be questionable that the accuracy suddenly dropped despite using the same training data set. However, it is important to note that the grid search cross validation process was performed on a substantially lower number of iterations and data points (10 epochs vs 150 & 100 batch vs 150 intially) as it was computationally infeasible due to hardware limitations.

Testing this parameter on the final test set also gave a similar accuracy score of 0.93. Objectively, an accuracy score of 0.93 is relatively high which shows that the model is able to generalise data sets and predict a malignant tumour to a passable level.It's also good to note that the training accuracy with GridSearchCV vs testing accuracy was extremely similar which indicates that overfitting isn't very present in our model (possibly largely due to the drop_out of neurons in the 2 hidden layers).

The f1 score shows a positive score of 0.88 when predicting malignant cases which means that the model is quite strong and is able to confidentally predict diagnosis correctly over all. The highlight of this model is that it has quite a high precision accuracy meaning that most of the samples that were predicted as Malignant were indeed Malignant. However, it is crucial to note that in our case study, a very high recall score is ideal so that we are able to avoid any false negatives as wrongly predicting a patient as benign is more costly to the hospital than wrongly predicting someone as malignant. In this scenario, those 5 false negatives can potentially cost the hospital several law suits and will not be acceptable for a life threatening situation such as cancer.

The model we fitted at the very end seems better at predicting whether a tumour is not malignant however it is good to note that the number of false positives and negatives are the same (and predicting benign cases is not the main purpose of this algorithm). In conclusion, the model needs to be tweaked a lot more before it can be deployed into the hospital as the recall score is not at a level that is acceptable. Further tuning can be performed (with better hardware improvements and laxed time constraints) with a higher number of iterations and/or datapoints by increasing the CV, epoch or batch parameters (these parameters can also be tuned for optimal training approaches). A parameter that could be tweaked to improve the cross validation process would be "n_jobs", setting it to -1 will quicken the process as it uses all cores of the GPU. Another potentially good parameter to look into would be "weight_constraint" which sets an upper bound to the weights and is used to spread out the weight values to improve model generalisation as well as decrease overfitting. Also the data set ideally can be increased to allow for better generalisation of future cases (around 100 currently isn't very large).

#####################################################################################################################